### PR TITLE
Validation of Numbers

### PR DESCRIPTION
**Pull request by TomiwaOke**
- Change region code to country code to allow for easy calling of metadata zip file ranges

- Addition of number validation process after migration

- Add Migration Report inner class with details about a given migration as well as export methods

- Including "data" folder containing metadata and recipes.csv in pom.xml so that test & tool runs properly from IDE and terminal (**Note:** the data inside the data file has not been pushed up yet because the location of the file is to be determined)

### CODE REVIEW

File: migrator/src/main/java/com/google/phonenumbers/migrator/CommandLineMain.java [hunk 1]
```diff
@@ -68,25 +66,52 @@ public static void main(String[] args) throws IOException {
...
+  }
+
+  private static void printNumberReport(MigrationReport mr) {
+    if (mr.getValidMigrations().size() == 1) {
+      System.out.println("Successful migration into: +"
+          + mr.getValidMigrations().get(0).getMigratedNumber());
+    } else if (mr.getInvalidMigrations().size() == 1) {
+      System.out.println("The number was migrated into '+"
+          + mr.getInvalidMigrations().get(0).getMigratedNumber() + "' but this number was not "
+          + "seen as being valid and dialable when inspecting our data for the given country");
```
Comment (brittanyroesch) [inline migrator/src/main/java/com/google/phonenumbers/migrator/CommandLineMain.java:110]: Could we mention the country they passed in for these print statements, just as a sanity check for the user.

File: migrator/src/test/java/com/google/phonenumbers/migrator/MigrationJobTest.java [hunk 2]
```diff
@@ -133,20 +139,58 @@ public void performMultipleMigration_nextRecipeNotFound_expectException() throws
...
+
+    assertThat(migratedNums.getValidMigrations().stream().map(MigrationResult::getMigratedNumber)
         .collect(Collectors.toList()))
         .containsExactly(DigitSequence.of(migratedNumber));
   }
+
+  @Test
+  public void standardMigration_invalidNumberNoRecipe_expectNoMigration() throws IOException {
+    String invalidNumber = "1234567";
+    String countryCode = "44";
```
Comment (brittanyroesch) [inline migrator/src/test/java/com/google/phonenumbers/migrator/MigrationJobTest.java]: I think a lot of these variables could be class level variables (ie. private static final COUNTRY_CODE = "44";)

File: migrator/src/test/java/com/google/phonenumbers/migrator/MigrationJobTest.java [hunk 3]
```diff
@@ -133,20 +139,58 @@ public void performMultipleMigration_nextRecipeNotFound_expectException() throws
...
+  @Test
+  public void standardMigration_migratableNumber_expectMigration() throws IOException {
+    String alreadyValidNumber = "841201234567";
+    String countryCode = "84";
+    MigrationJob job = MigrationFactory.createMigration(alreadyValidNumber, countryCode);
+
+    MigrationReport report = job.getMigrationReportForCountry();
+    assertThat(report.getValidMigrations().stream().map(MigrationResult::getOriginalNumber))
+        .containsExactly(alreadyValidNumber);
+  }
```
Comment (brittanyroesch) [inline migrator/src/test/java/com/google/phonenumbers/migrator/MigrationJobTest.java:195]: Is it possible to test a migration that became invalid afterwards?

**Review by brittanyroesch — COMMENTED** [review]
Great job! Just a few small comments :)

File: migrator/src/main/java/com/google/phonenumbers/migrator/CommandLineMain.java (hunk 1 above)
Comment (TomiwaOke) [reply migrator/src/main/java/com/google/phonenumbers/migrator/CommandLineMain.java:110]: Country code print out added above in main method

**Review by brittanyroesch — APPROVED** [review]
LGTM - Thanks!

File: migrator/src/main/java/com/google/phonenumbers/migrator/MigrationJob.java [hunk 4]
```diff
@@ -167,4 +206,155 @@ private DigitSequence getMigratedValue(String staleString, String oldFormat, Str
...
+    private final ImmutableList<MigrationResult> invalidMigrations;
+
+    private MigrationReport(Stream<MigrationEntry> untouchedEntries,
+        ImmutableMap<String, ImmutableList<MigrationResult>> migratedEntries) {
+      this.untouchedEntries = untouchedEntries.collect(ImmutableList.toImmutableList());
+      this.validMigrations = migratedEntries.get("Valid");
+      this.invalidMigrations = migratedEntries.get("Invalid");
+    }
+
+    public ImmutableList<MigrationResult> getValidMigrations() {
```
Comment (gonecoding) [inline migrator/src/main/java/com/google/phonenumbers/migrator/MigrationJob.java:260]: I believe this and the subsequent two public functions could use a one-liner of JavaDoc explaining what exactly they return.

File: migrator/src/main/java/com/google/phonenumbers/migrator/MigrationJob.java [hunk 5]
```diff
@@ -167,4 +206,155 @@ private DigitSequence getMigratedValue(String staleString, String oldFormat, Str
...
+      return untouchedEntries;
+    }
+
+    /**
+     * Creates a text file of the new number list after a migration has been performed. Numbers that
+     * were not migrated are added in their original format as well migrated numbers that were seen
+     * as being invalid based on the given rangesTable for the given countryCode. Successfully
+     * migrated numbers will be added in their new format.
+     */
+    public String exportToFile(String originalFileName) throws IOException {
```
Comment (gonecoding) [inline migrator/src/main/java/com/google/phonenumbers/migrator/MigrationJob.java]: Add JavaDoc for parameter?

File: migrator/src/main/java/com/google/phonenumbers/migrator/MigrationUtils.java [hunk 6]
```diff
@@ -41,47 +39,47 @@
...
 
-    return migrationEntries.stream()
+    return migrationEntries
         .filter(entry -> recipeKey.asRangeTree().contains(entry.getSanitizedNumber()));
   }
 
   /**
-   * Returns the sub range of entries within migrationEntries that can be migrated using any recipe
-   * from the {@link CsvTable} recipesTable that matches the specified BCP-47 region code. This
+   * Returns the sub range of entries within migrationEntires that can be migrated using any recipe
```
Comment (gonecoding) [inline migrator/src/main/java/com/google/phonenumbers/migrator/MigrationUtils.java]: Typo? migrationEntires

**Review by gonecoding — APPROVED** [review]

**Pull request merged**